### PR TITLE
Enhance localization of prepareSecondsForDisplay (#2588)

### PR DESCRIPTION
**Pull request by johnPhillips**
## Summary

Addresses issue #2588

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

File: assets/js/util/i18n.js [hunk 1]
```diff
@@ -33,28 +33,51 @@ import { __, sprintf } from '@wordpress/i18n';
...
+ * @param {(Intl.NumberFormatOptions)} [options] Optional formatting options.
+ * @return {string}   		Human readable string indicating time elapsed.
  */
-export const prepareSecondsForDisplay = ( seconds ) => {
+const durationFormat = ( seconds, options = { unitDisplay: 'short' } ) => {
 	seconds = parseInt( seconds, 10 );
 
-	if ( isNaN( seconds ) || 0 === seconds ) {
-		return '0.0s';
+	if ( isNaN( seconds ) ) {
```
Comment (eugene-manuilov) [inline assets/js/util/i18n.js:46]: We should use `Number.isNaN` isntead of `isNaN`:

```suggestion
	if ( Number.isNaN( seconds ) ) {
```

File: assets/js/util/i18n.js [hunk 2]
```diff
@@ -33,28 +33,51 @@ import { __, sprintf } from '@wordpress/i18n';
...
+	}
+
+	seconds = numberFormat( seconds, {
+		style: 'unit',
+		unit: 'second',
+		...options,
+	} );
+
+	/* translators: 1: hours, 2: minutes, 3: seconds */
+	const formattedString = sprintf( _x( '%1$s %2$s %3$s', 'google-site-kit', 'google-site-kit' ), hours, minutes, seconds );
```
Comment (eugene-manuilov) [inline assets/js/util/i18n.js:78]:
```suggestion
	const formattedString = sprintf( _x( '%1$s %2$s %3$s', 'duration of time: hh mm ss', 'google-site-kit' ), hours, minutes, seconds );
```

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Overall looks good to me. Just added two minor comments to address. Once fixed it should be good to go.

**Review by eugene-manuilov — APPROVED** [review]

File: assets/js/util/i18n.js [hunk 3]
```diff
@@ -33,28 +33,51 @@ import { __, sprintf } from '@wordpress/i18n';
...
+	let hours = Math.floor( seconds / 60 / 60 ) || '';
+	let minutes = Math.floor( ( seconds / 60 ) % 60 ) || '';
+
+	seconds = Math.floor( seconds % 60 );
+
+	if ( hours ) {
+		hours = numberFormat( hours, {
+			style: 'unit',
+			unit: 'hour',
+			...options,
```
Comment (aaemnnosttv) [inline assets/js/util/i18n.js:59]: Let's move where the spread happens since the `style` and `unit` shouldn't be overridable.

File: assets/js/util/i18n.js [hunk 4]
```diff
@@ -33,28 +33,51 @@ import { __, sprintf } from '@wordpress/i18n';
...
+	}
+
+	seconds = numberFormat( seconds, {
+		style: 'unit',
+		unit: 'second',
+		...options,
+	} );
+
+	/* translators: 1: hours, 2: minutes, 3: seconds */
+	const formattedString = sprintf( _x( '%1$s %2$s %3$s', 'duration of time: hh mm ss', 'google-site-kit' ), hours, minutes, seconds );
```
Comment (aaemnnosttv) [inline assets/js/util/i18n.js:78]: Let's split this onto a few lines as `sprintf` gets a bit hard to follow with the translated string and multiple parameters all on one line. This should do it (although it gets confused about the indentation of the translators comment sometimes 😄 )
```suggestion
	const formattedString = sprintf(
		/* translators: 1: hours, 2: minutes, 3: seconds */
		_x( '%1$s %2$s %3$s', 'duration of time: hh mm ss', 'google-site-kit' ),
		hours,
		minutes,
		seconds
	);
```

One additional thought here – perhaps we should document the placeholders a bit more descriptively since each one will also have its localized units included (if non-zero), so we should make it clear to avoid translators adding unit prefix/suffixes.

File: assets/js/util/i18n.js [hunk 5]
```diff
@@ -33,28 +33,51 @@ import { __, sprintf } from '@wordpress/i18n';
...
  *
- * @param {number} seconds The number of seconds.
- * @return {string} Human readable string indicating time elapsed.
- *
+ * @param {number}                     seconds   The number of seconds.
+ * @param {(Intl.NumberFormatOptions)} [options] Optional formatting options.
+ * @return {string}   		Human readable string indicating time elapsed.
  */
-export const prepareSecondsForDisplay = ( seconds ) => {
+const durationFormat = ( seconds, options = { unitDisplay: 'short' } ) => {
```
Comment (aaemnnosttv) [inline assets/js/util/i18n.js:43]: Generally we always apply defaults in the destructuring, not in the default value itself. That way the individual defaults aren't lost if `options` are passed.

```suggestion
const durationFormat = ( seconds, { unitDisplay = 'short' } = {} ) => {
```

In this case, since we want all of the options in the one object it makes more sense to handle the defaults in the function, then we don't need to worry about passing `options.style` as well.

```js
const durationFormat = ( seconds, options = {} ) => {
	const options = {
		unitDisplay: 'short',
		...options,
		style: 'unit',
	};

```

File: assets/js/util/i18n.js [hunk 6]
```diff
@@ -180,10 +203,9 @@ export const numFmt = ( number, options = {} ) => {
 
 	if ( 'metric' === style ) {
 		return readableLargeNumber( number );
-	}
-
-	if ( 'duration' === style ) {
-		return prepareSecondsForDisplay( number );
+	} else if ( 'duration' === style ) {
+		formatOptions = omit( options, 'style' );
```
Comment (aaemnnosttv) [inline assets/js/util/i18n.js:207]: Let's remove the `omit` here – `durationFormat` should simply ignore/override it if passed.

File: assets/js/util/i18n.js (hunk 4 above)
Comment (aaemnnosttv) [inline assets/js/util/i18n.js:78]: Looking at this again and thinking longer term, it seems reasonable that we might extend this to include larger units of time in the future (e.g. days, weeks...) at which point adding these placeholders would be a bit awkward.

Let's reverse the order so that `%1$s` refers to seconds, `%2$s` to minutes, and `%3$s` to hours so that the number of the placeholder increases with the unit of time.

**Review by aaemnnosttv — CHANGES_REQUESTED** [review]
Thanks @johnPhillips – this looks great, just a few small things to address and then I think this should be good to go.

File: assets/js/util/i18n.js (hunk 4 above)
Comment (johnPhillips) [reply assets/js/util/i18n.js:78]: @aaemnnosttv Something like `duration of time (numeric): hh mm ss` perhaps, or does it need to be more verbose than that?

File: assets/js/util/i18n.js (hunk 3 above)
Comment (johnPhillips) [reply assets/js/util/i18n.js:59]: Good catch

File: assets/js/util/i18n.js (hunk 4 above)
Comment (johnPhillips) [reply assets/js/util/i18n.js:78]: Good idea

File: assets/js/util/i18n.js [hunk 7]
```diff
@@ -33,28 +33,62 @@ import { __, sprintf } from '@wordpress/i18n';
...
+
+	let hours = Math.floor( seconds / 60 / 60 ) || '';
+	let minutes = Math.floor( ( seconds / 60 ) % 60 ) || '';
+
+	seconds = Math.floor( seconds % 60 );
+
+	if ( hours ) {
+		hours = numberFormat( hours, {
+			...options,
+			style: 'unit',
```
Comment (aaemnnosttv) [inline assets/js/util/i18n.js:64]: We can remove the `style` in these calls since it's applied to all options now above.

File: assets/js/util/i18n.js [hunk 8]
```diff
@@ -33,28 +33,62 @@ import { __, sprintf } from '@wordpress/i18n';
...
+		style: 'unit',
+		unit: 'second',
+	} );
+
+	const formattedString = sprintf(
+		/* translators: 1: seconds, 2: minutes, 3: hours */
+		_x( '%1$s %2$s %3$s', 'duration of time (numeric): hh mm ss', 'google-site-kit' ),
+		hours,
+		minutes,
+		seconds
```
Comment (aaemnnosttv) [inline assets/js/util/i18n.js:88]: The resulting `formattedString` will be correct for the default locale here but `hours` should refer to `%3$s` where it is currently filling the `%1$s` placeholder. We just need to reverse the order of the placeholders in the string and the order they're passed to `sprintf` which should result in the same output but fix the relationship between placeholders and inputs. The translators comment is correct although it might be more clear to emphasize that each unit is formatted (rather than a bare number).

File: assets/js/util/test/prepareSecondsForDisplay.js [hunk 9]
```diff
@@ -1,41 +0,0 @@
...
-		'1m 5s',
-	],
-	[
-		( 60 * 60 * 3 ) + ( 60 * 5 ) + 12,
-		'3h 5m 12s',
-	],
-	[
-		( 60 * 60 * 7 ) + ( 60 * 2 ) + 42,
-		'7h 2m 42s',
-	],
```
Comment (aaemnnosttv) [inline assets/js/util/test/prepareSecondsForDisplay.js:34]: Let's also include/merge these cases with the new tests. There are a few cases here which are important to keep such as `35`, `60` and `65` (although we don't need to include `65` twice like it is here 😄 ). It would also be good to ensure we're not breaking anything in the process.

File: assets/js/util/i18n.js [hunk 10]
```diff
@@ -33,28 +33,62 @@ import { __, sprintf } from '@wordpress/i18n';
...
  * @since 1.0.0
+ * @since n.e.x.t Refactored and renamed to improve localization.
+ * @private
  *
- * @param {number} seconds The number of seconds.
- * @return {string} Human readable string indicating time elapsed.
- *
+ * @param {number}                     seconds   The number of seconds.
+ * @param {(Intl.NumberFormatOptions)} [options] Optional formatting options.
+ * @return {string}   		Human readable string indicating time elapsed.
```
Comment (aaemnnosttv) [inline assets/js/util/i18n.js:41]: The return description doesn't require any special indentation, only parameters need to be aligned.

**Review by aaemnnosttv — COMMENTED** [review]
Almost there!

File: assets/js/util/i18n.js (hunk 8 above)
Comment (johnPhillips) [reply assets/js/util/i18n.js:88]: Sorry, I think I misunderstood this last time round. Hopefully fixed now.

**Comment by aaemnnosttv** [comment]
@googlebot I consent.

**Review by aaemnnosttv — APPROVED** [review]
Super, thanks!
